File: backend/src/main/java/com/etekhno/rentacar/domain/PaymentDTO.java
package com.etekhno.rentacar.domain;

import com.etekhno.rentacar.datamodel.enums.PaymentStatus;
import com.etekhno.rentacar.domain.inbound.PaymentDTOIn;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.util.Date;

@Setter
@Getter
@NoArgsConstructor
public class PaymentDTO extends PaymentDTOIn {
    private String id;
    private BigDecimal overpaidAmount = BigDecimal.ZERO;
    private Date paymentDate;
    private PaymentStatus paymentStatus;

    public PaymentDTO(String id, String customerId, BigDecimal totalAmount, String paymentMethod,
                      BigDecimal overpaidAmount, Date paymentDate, PaymentStatus paymentStatus) {
        super(customerId, totalAmount, paymentMethod);
        this.id = id;
        this.overpaidAmount = overpaidAmount;
        this.paymentDate = paymentDate;
        this.paymentStatus = paymentStatus;
    }
}

File: frontend/lib/domains/payment_dto.dart
import '/domains/enums/payment_status.dart';
import '/domains/inbound/payment_dto_in.dart';

class PaymentDTO extends PaymentDTOIn {
  String? id;
  double? overpaidAmount = 0;
  DateTime? paymentDate = DateTime.now();
  PaymentStatus? paymentStatus;

  PaymentDTO({
    required this.id,
    this.overpaidAmount,
    this.paymentDate,
    this.paymentStatus,
    required super.customerId,
    required super.totalAmount,
    required super.paymentMethod,
  });

  factory PaymentDTO.fromJson(Map<String, dynamic> json) {
    return PaymentDTO(
      id: json["id"] as String,
      overpaidAmount: (json["overpaidAmount"] as num).toDouble(),
      paymentDate: DateTime.tryParse(json["paymentDate"] as String),
      paymentStatus: _parsePaymentStatus(json["paymentStatus"] as String),
      customerId: json["customerId"] as String,
      totalAmount: (json["totalAmount"] as num).toDouble(),
      paymentMethod: json["paymentMethod"],
    );
  }

  PaymentStatus parsePaymentStatus(String paymentStatus) {
    return _parsePaymentStatus(paymentStatus);
  }

  static PaymentStatus _parsePaymentStatus(String paymentStatus) {
    switch (paymentStatus) {
      case 'COMPLETED':
        return PaymentStatus.COMPLETED;
      case 'FAILED':
        return PaymentStatus.FAILED;
      default:
        return PaymentStatus.UNALLOCATED;
    }
  }
}
